crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -66,6 +88,75 @@ impl<'db> UnionBuilder<'db> {
             }
             // Adding `Never` to a union is a no-op.
             Type::Never => {}
+            // If adding a string literal, look for an existing `UnionElement::StringLiterals` to
+            // add it to, or an existing element that is a super-type of string literals, which
+            // means we shouldn't add it. Otherwise, add a new `UnionElement::StringLiterals`
+            // containing it.
+            Type::StringLiteral(literal) => {
+                let mut found = false;
+                for element in &mut self.elements {
+                    match element {
+                        UnionElement::StringLiterals(literals) => {
+                            literals.insert(literal);
+                            found = true;
+                            break;
+                        }
+                        UnionElement::Type(existing) if ty.is_subtype_of(self.db, *existing) => {
+                            return self;
+                        }
+                        _ => {}
+                    }
+                }
+                if !found {
+                    let mut literals = FxOrderSet::default();
+                    literals.insert(literal);
+                    self.elements.push(UnionElement::StringLiterals(literals));
```

Comment:
Slightly disappointing that we can't use something like the following here (you might have tried as well). Doesn't work with the `FxHasher`, for some reason.
```suggestion
                    self.elements.push(UnionElement::StringLiterals([literal].into()));
```

Reply:
This _does_ compile, however:

```suggestion
                    self.elements
                        .push(UnionElement::StringLiterals(FxOrderSet::from_iter([
                            literal,
                        ])));
```